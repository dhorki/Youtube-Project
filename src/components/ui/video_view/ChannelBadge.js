import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledChannelBadge } from '../../../styles/components/ui/video_view/ChannelBadge';
import { useFetch } from '../../../hooks/useFetch';
import { intToMagnitude } from '../../../helpers/formatters';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const ChannelBadge = ({ id }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/channels?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet,statistics&' +
    `id=${id}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);

  let image = '',
    title = '#?',
    subscriberCount = 0;
  if (!loading && !error) {
    const item = data?.items[0];
    image = item.snippet.thumbnails.default.url;
    title = item.snippet.title;
    subscriberCount = item.statistics.subscriberCount;
  }

  return (
    <StyledChannelBadge theme={theme}>
      {loading ? null : (
        <div className="badge">
          {!error ? <img src={image} alt={title} /> : null}
          <div>
            <p className="title">{title}</p>
            <p className="subscribers">{intToMagnitude(subscriberCount)} subscribers</p>
          </div>
        </div>
      )}
    </StyledChannelBadge>
  );
};

ChannelBadge.propTypes = {
  id: PropTypes.string.isRequired,
};
