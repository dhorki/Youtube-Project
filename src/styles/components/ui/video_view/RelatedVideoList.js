import styled from 'styled-components';
import { screenSizes } from '../../../../constants/constants';

export const StyledRelatedVideoList = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 390px;

  @media only screen and (max-width: ${screenSizes.mid}) {
    width: 100%;
    flex-direction: row;
    align-items: top;
  }

  > h3 {
    color: ${({ theme }) => theme.primary};
    width: 100%;
    text-align: center;
  }
`;
