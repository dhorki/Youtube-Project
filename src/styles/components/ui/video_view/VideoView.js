import styled from 'styled-components';
import { screenSizes, styles } from '../../../../constants/constants';

const videoArea = '60vw';
const videoAreaSmall = '100vw';

export const StyledVideoView = styled.div`
  max-width: ${videoArea};
  min-width: 480px;

  @media only screen and (max-width: ${screenSizes.mid}) {
    max-width: ${videoAreaSmall};
  }

  .embedded-video {
    width: 100%;
    min-width: ${screenSizes.xsmall};
    min-height: calc(480px * 9 / 16);
    height: calc(${videoArea} * 9 / 16);

    @media only screen and (max-width: ${screenSizes.mid}) {
      height: calc(${videoAreaSmall} * 9 / 16);
    }

    iframe {
      width: 100%;
      height: 100%;
      border: none;
    }
  }
  > .view-video-title {
    color: ${({ theme }) => theme.primary};
    font-size: 20px;
    font-weight: bolder;
    margin: 15px 0px;
    padding: 0;
    text-align: left;
    width: 100%;
    min-width: 480px;
  }

  > .view-video-stats {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    color: ${({ theme }) => theme.secondary};
    font-size: 13px;
    font-weight: bolder;
    margin: 0px;
    text-align: left;
    width: 100%;
    min-width: 480px;

    > p {
      margin: 0px;
    }
    > .view-video-stats-right {
      font-size: 15px;

      .live-icon {
        color: ${({ theme }) => theme.appIcon};
      }
    }
  }

  > .view-video-description-box {
    display: flex;
    flex-direction: column;

    > .view-video-description {
      color: ${({ theme }) => theme.secondary};
      overflow: hidden;
      min-height: 100px;
      height: 100%;
    }

    > .view-video-description.collapsed {
      height: 100px;
    }

    > .view-video-description-toggle {
      color: ${({ theme }) => theme.secondary};
      font-weight: bolder;
      background-color: transparent;
      border: none;
      justify-content: left;
      text-align: left;
      margin-top: 30px;
      width: 100px;
      cursor: pointer;
    }
  }

  hr {
    width: 100%;
    border: none;
    border-top: 1px solid ${({ theme }) => theme.inputBorder};
    margin: 15px 0px;
  }

  > .view-video-channel-favorite {
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;

    > button {
      width: 180px;
      height: 40px;
      color: ${styles.colors.dark.primary};
      background-color: ${styles.colors.light.appIcon};
      font-weight: bolder;
      border-radius: 5px;
      border: none;
      cursor: pointer;

      &.remove {
        color: ${styles.colors.dark.primary};
        background-color: ${styles.colors.light.googleBlue};
      }
    }
  }
`;
