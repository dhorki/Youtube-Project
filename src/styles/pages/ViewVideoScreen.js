import styled from 'styled-components';
import { screenSizes } from '../../constants/constants';

export const StyledViewVideoScreen = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;

  @media only screen and (max-width: ${screenSizes.mid}) {
    flex-direction: column;
  }

  margin: 40px;
  margin-top: 15px;
`;
