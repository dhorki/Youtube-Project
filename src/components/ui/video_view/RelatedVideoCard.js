import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { timeSince } from '../../../helpers/formatters';
import { decodeHtml } from '../../../helpers/decodeHtml';
import { StyledRelatedVideoCard } from '../../../styles/components/ui/video_view/RelatedVideoCard';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoCard = ({ video }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const { id, snippet } = video;
  const { videoId } = id;
  const { title, channelTitle, thumbnails, publishTime } = snippet;
  const { url } = thumbnails.medium;
  const d = new Date(publishTime);

  const upperCaseCharsCount = title.length - title.replaceAll(/[A-Z]/g, '').length;
  const limitLength = upperCaseCharsCount > 50 ? 50 : upperCaseCharsCount > 30 ? 60 : 70;

  const fixedTitle =
    title.length >= limitLength ? title.slice(0, limitLength) + '...' : title;

  return (
    <StyledRelatedVideoCard
      className="animate__animated animate__fadeIn"
      theme={theme}
      to={`/${videoId}`}
    >
      <img src={url} alt="video" />
      <div>
        <div className="videocard-title">{decodeHtml(fixedTitle)}</div>
        <div className="videocard-channel-title">
          {decodeHtml(channelTitle)}
          <br />
          {timeSince(d)} ago.
        </div>
      </div>
    </StyledRelatedVideoCard>
  );
};

RelatedVideoCard.propTypes = {
  video: PropTypes.object.isRequired,
};
