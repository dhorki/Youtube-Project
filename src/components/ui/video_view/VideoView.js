import React, { useState } from 'react';
import { RiLiveLine } from 'react-icons/ri';
import { FaThumbsUp } from 'react-icons/fa';
import { IoTimerOutline } from 'react-icons/io5';
import {
  formatYoutubeDate,
  intToMagnitude,
  numberWithCommas,
  youtubeDurationToTime,
} from '../../../helpers/formatters';
import PropTypes from 'prop-types';
import { useFetchYoutubeVideos } from '../../../hooks/useFetchYoutubeVideos';
import { useTheme } from '../../../hooks/useTheme';

import { StyledVideoView } from '../../../styles/components/ui/video_view/VideoView';
import { LoadingAnimation } from '../loading/LoadingAnimation';
import { ChannelBadge } from './ChannelBadge';
import { ErrorText } from '../error/ErrorText';

export const VideoView = ({ id }) => {
  const { getTheme } = useTheme();
  const theme = getTheme();

  const { data, loading, error } = useFetchYoutubeVideos(
    process.env.REACT_APP_YOUTUBE_API_KEY,
    [id]
  );

  const [descriptionCollapsed, setDescriptionCollapsed] = useState(true);

  const iframeAllow =
    'accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture';

  // const { items } = require('../../../mocks/youtube-video-mock.json');
  let snippet, contentDetails, statistics;
  if (!loading && !error) {
    console.log(data);
    const item = data?.items[0];
    snippet = item.snippet;
    contentDetails = item.contentDetails;
    statistics = item.statistics;
  }

  return (
    <StyledVideoView className="animate__animated animate__fadeIn" theme={theme}>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : (
        <>
          <div className="embedded-video">
            <iframe
              title={id}
              src={`//www.youtube.com/embed/${id}`}
              frameBorder="0"
              allow={iframeAllow}
              allowFullScreen="allowfullscreen"
            />
          </div>
          <p className="view-video-title">{snippet.title}</p>
          <div className="view-video-stats">
            <p className="view-video-stats-left">
              <span>{numberWithCommas(statistics.viewCount)}</span> views -{' '}
              <span>{formatYoutubeDate(snippet.publishedAt)}</span>
            </p>
            <p className="view-video-stats-right">
              <span>
                <FaThumbsUp /> {intToMagnitude(statistics.likeCount)}
              </span>{' '}
              <span>
                <IoTimerOutline />{' '}
                {snippet.liveBroadcastContent === 'live' ? (
                  <RiLiveLine className="live-icon" />
                ) : (
                  youtubeDurationToTime(contentDetails.duration)
                )}
              </span>
            </p>
          </div>
          <hr />
          <div className="view-video-channel-favorite">
            <ChannelBadge id={snippet.channelId} />
            <button>ADD TO FAVORITES</button>
          </div>
          <div className="view-video-description-box">
            <p
              className={`view-video-description ${
                descriptionCollapsed ? 'collapsed' : ''
              }`}
              dangerouslySetInnerHTML={{
                __html: snippet.description.replaceAll('\n', '<br />'),
              }}
            />
            <button
              className="view-video-description-toggle"
              onClick={() => setDescriptionCollapsed(!descriptionCollapsed)}
            >
              SHOW {descriptionCollapsed ? 'MORE' : 'LESS'}
            </button>
          </div>
        </>
      )}
    </StyledVideoView>
  );
};

VideoView.propTypes = {
  id: PropTypes.string.isRequired,
};
