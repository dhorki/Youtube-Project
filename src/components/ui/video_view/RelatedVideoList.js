import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledRelatedVideoList } from '../../../styles/components/ui/video_view/RelatedVideoList';
import { useFetch } from '../../../hooks/useFetch';
import { RelatedVideoCard } from './RelatedVideoCard';
import { LoadingAnimation } from '../loading/LoadingAnimation';
import { ErrorText } from '../error/ErrorText';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoList = ({ relatedTo }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/search?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet&' +
    'type=video&' +
    'order=relevance&' +
    `relatedToVideoId=${relatedTo}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);
  // const { data, loading, error } = { data: [], loading: true, error: null };

  let filteredItems = [];

  if (!loading && !error) {
    filteredItems = data.items?.filter(
      (item) => item.id.kind.includes('youtube#video') && item.snippet
    );
  }

  return (
    <StyledRelatedVideoList theme={theme}>
      <h3>Related Videos</h3>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : !filteredItems || filteredItems.length === 0 ? (
        'No Related videos.'
      ) : (
        filteredItems.map((video) => (
          <RelatedVideoCard key={video.id.videoId} video={video} />
        ))
      )}
    </StyledRelatedVideoList>
  );
};

RelatedVideoList.propTypes = {
  relatedTo: PropTypes.string.isRequired,
};
