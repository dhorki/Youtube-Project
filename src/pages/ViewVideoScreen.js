import React from 'react';
import { useParams } from 'react-router-dom';
import { RelatedVideoList } from '../components/ui/video_view/RelatedVideoList';
import { VideoView } from '../components/ui/video_view/VideoView';
import { StyledViewVideoScreen } from '../styles/pages/ViewVideoScreen';

export const ViewVideoScreen = () => {
  const { videoId } = useParams();
  return (
    <StyledViewVideoScreen>
      <VideoView id={videoId} />
      <RelatedVideoList relatedTo={videoId} />
    </StyledViewVideoScreen>
  );
};
