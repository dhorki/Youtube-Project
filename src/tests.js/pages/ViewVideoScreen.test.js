import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { ViewVideoScreen } from '../../pages/ViewVideoScreen';
import { MemoryRouter } from 'react-router-dom';
import { VideoView } from '../../components/ui/video_view/VideoView';
import { RelatedVideoList } from '../../components/ui/video_view/RelatedVideoList';
import { Route } from 'react-router-dom';
jest.mock('../../components/ui/video_view/VideoView');
jest.mock('../../components/ui/video_view/RelatedVideoList');

describe('VideoViewScreen', () => {
  it('should render correctly', () => {
    const mockLocation = {
      pathname: '/mockVideoId',
      search: '',
    };

    VideoView.mockImplementation(({ id }) => <div>VideoView {id}</div>);
    RelatedVideoList.mockImplementation(({ relatedTo }) => (
      <div>RelatedVideoList {relatedTo}</div>
    ));

    const wrapper = mount(
      <MemoryRouter initialEntries={[mockLocation]}>
        <Route path="/:videoId">
          <ViewVideoScreen />
        </Route>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
