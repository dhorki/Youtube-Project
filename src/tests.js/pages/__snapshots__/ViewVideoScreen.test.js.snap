// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoViewScreen should render correctly 1`] = `
<div
  className="sc-bczRLJ dhxzA-D"
>
  <div>
    VideoView 
    mockVideoId
  </div>
  <div>
    RelatedVideoList 
    mockVideoId
  </div>
</div>
`;
